### DISCUSSION TITLE
prepINITModel not ignoring custom reactions in model pruning

### DISCUSSION THREAD

**Discussion by wshao1** [SysBioChalmers/RAVEN#507] [Help me!] [answered]
Hi! I am trying to apply ftINIT to only a subset of the Human1 model, by adding the reactions I want to ignore to the `customRxnsToIgnore` argument of `prepINITModel`. However, after running ftINIT with and without my customRxnsToIgnore, I see no difference in reaction inclusion. What might be causing this? Am I misinterpreting how to use customRxnsToIgnore?

**Comment by mihai-sysbio** [comment]
Hi @wshao1! Just judging by the [documentation](https://sysbiochalmers.github.io/RAVEN/doc/INIT/prepINITModel.html), I too find it unclear. I'm wondering if the explanation could be more simple, e.g. the reaction identifier is not in the right format. Could you share a bit of the sample code?

**Reply by wshao1** [reply]
Yes! As a bit of background, I am trying to create a GEM for the HCT116 cancer cell line, while protecting the reactions associated with essential genes. My code below is based on the [ftINIT guide](https://sysbiochalmers.github.io/Human-GEM-guide/gem_extraction/).

First I prepare Human1 for ftINIT by running a modified version of `prepHumanModelForftINIT` ([prepHumanModelForftINIT.txt](https://github.com/SysBioChalmers/RAVEN/files/12313472/prepHumanModelForftINIT.txt)), where I add an additional argument `essentialRxns` to append reactions to `customRxnsToIgnore`.

```
prepDataHCT116 = prepHumanModelForftINIT(ihuman, false, human_tasks_path, human1_rxns_path, essentialRxns)
save('prepDataHCT116.mat', 'prepDataHCT116')
```

Then I proceed to run ftINIT using the resulting prepData structure.

```
%% Load DepMap transcriptomics data
hct116_tpms = readtable('../raw_data/hct116_tpms/hct116_tpms.csv')

%% Convert data into input structure
hct116_struct.genes = hct116_tpms{:, 1};    % gene names
hct116_struct.tissues = {'HCT116'};    % sample name
hct116_struct.levels = hct116_tpms{:, 2};    % gene TPM values
hct116_struct.threshold = 0.02;

%% Run ftINIT
hct116_model = ftINIT(prepDataHCT116, hct116_struct.tissues{1}, [], [], hct116_struct, {}, getHumanGEMINITSteps('1+1'), true, true);
```

I don't believe the issue stems from incorrect reaction identifiers, because I do see more reactions marked with a 1 in the `prepData.toIgnoreCustomRxns` field. The issue seems to be how ftINIT processes the `toIgnoreCustomRxns` field to protect those reactions from removal.

**Reply by mihai-sysbio** [reply]
Having had a deeper look, I'm actually confused that as params `'1+1'` is passed for the steps, which is the same as the [default option](https://github.com/SysBioChalmers/RAVEN/blob/da86f6246b27b1d347850375d718933842077732/INIT/ftINIT.m#L103-L104) but then that might not fully correspond with the steps that are required to ignore the custom reactions https://github.com/SysBioChalmers/RAVEN/blob/da86f6246b27b1d347850375d718933842077732/INIT/ftINIT.m#L168-L183

@johan-gson could you have a look at this?

**Comment by johan-gson** [comment, marked as answer]
Hi @wshao1 , customRxnsToIgnore does not do this, these are the reactions that are not included in the first step in the algorithm, so something different. I have not thought of this case, there is no direct support for it, it would eventually be a good thing to add as it would be a very simple fix. I'm very short on time right now, so I cannot engage in any development right now. If you want, you could modify the code to take care of this. A suggestion is to test the following: In prepINITModel, add two parameters:
function prepData = prepINITModel(origRefModel, taskStruct, spontRxnNames, convertGenes, customRxnsToIgnore, extComp, skipScaling, extraEssentialRxns, extraEssentialDirs)

you send in these extra reaction ids when you call the prepINITModel function. You need to decide on a direction in which these reactions are essential (you cannot easily do both). If irrev, then it is easy, always forward (1), otherwise it can also be backwards (-1)
ex.: 
extraEssentialRxns = {'MARXXXX';'MARYYYY'}; % a vertical cell array, not horizontal, the reaction ids of the extra essential reactions
extraEssentialDirs = [1;-1]; %vertical vector

Then, on line 155 or somewhere around there, add these lines:
%Flip extra essential reactions with reversed essential direction
minModel1 = reverseRxns(minModel1, extraEssentialRxns(extraEssentialDirs < 0));
%Make the extra essential reactions irrev
sel = ismember(minModel1.rxns, extraEssentialDirs);
minModel1.rev(sel) = 0;
minModel1.lb(sel) = 0;

i.e., insert afterthese lines:
    minModel1 = reverseRxns(minModel1, minModel1.rxns(selInd(essentialRevDir)));
    %constructEquations(minModel2,minModel2.rxns(selInd(essentialRevDir))) %looks good

    %Then make them irreversible
    minModel1.rev(selInd) = 0;
    minModel1.lb(selInd) = 0;

I think this should work, definitely worth a try! Let me know how it went if you try it! Note that the code does not mess up the model that you get out (making irreversible etc.), we are just changing the specialized model used in the MILP, so it is just internal for the algorithm!

Best of luck!

**Comment by Christoff1993** [comment]
Hi!

I was interested in doing the same thing. In my case, I wanted to have ftINIT prune away at all subsystems in Human-GEM except the central carbon metabolism (Glycolysis, PPP, TCA, OXPHOS, and Pyruvate metabolism). I also want to make sure all amino acid exchanges stay put. The result is a cell array of 174 reaction IDs. I selected the forward direction for all of them.

I have managed to edit prepINITModel according to your suggestion:

    %first flip the reactions with negative direction
    %constructEquations(minModel1,minModel1.rxns(selInd(essentialRevDir))) %looks good
    minModel1 = reverseRxns(minModel1, minModel1.rxns(selInd(essentialRevDir)));
    %constructEquations(minModel2,minModel2.rxns(selInd(essentialRevDir))) %looks good

    **%CUSTOM: Flip extra essential reactions with reversed essential direction
    minModel1 = reverseRxns(minModel1, extraEssentialRxns(extraEssentialDirs < 0));
    %CUSTOM: Make the extra essential reactions irrev
    sel = ismember(minModel1.rxns, extraEssentialRxns);
    minModel1.rev(sel) = 0;
    minModel1.lb(sel) = 0;**

    %Then make them irreversible
    minModel1.rev(selInd) = 0;
    minModel1.lb(selInd) = 0;

The function now also accepts these as inputs:
`prepData = prepINITModelCustom(origRefModel, taskStruct, spontRxnNames, convertGenes, customRxnsToIgnore, extComp, skipScaling, extraEssentialRxns, extraEssentialDirs)`

prepData is generated without a problem. However, then I pass this to the ftINIT function, as follows:

```
cellGEM = ftINIT(refModel, 'data', [], [], data_struct, [], getINITSteps([],'1+1'), ...
                                true, true, [], true);
```

I get the following error:

```
**_Explored 0 nodes (0 simplex iterations) in 0.08 seconds (0.01 work units)
Thread count was 1 (of 12 available processors)

Solution count 0

Model is infeasible
Best objective -, best bound -, gap -
Error using dispEM (line 49)
Failed to find good enough solution within the time frame. MIPGap: Inf


Error in ftINIT (line 283)
        dispEM(['Failed to find good enough solution within the time frame. MIPGap: ' num2str(mipGap)]);

Error in B_Making_ftINIT_models (line 102)
    cellGEM = ftINIT(refModel, cell, celltype, hpaData, transcrData, metabolomicsData, INITSteps, ..._**
```

I manage to run the ftINIT fine with prepData if I make is using prepHumanModelForftINIT. So gurobi seems to be working fine.

Can you think of any reasons why this would still not be working?
